docs/dev/designs/RestFrontEnd.md

```diff
@@ -0,0 +1,136 @@
+
+It is then possible to generate a token with an SQL command:
+
+```sql
+ALTER USER <user_name> CREATE TOKEN '<token_name>';
+```
+
+It is possible to delete a token will this SQL commands:
+
+```sql
+ALTER USER <user_name> DROP TOKEN '<token_name>';
+```
+
+A user must provide the Siodb user name plus a valid token to access and manipulate the data:
+
+```curl --user "Siodb_user_name"  https://siodb-srv:80/<path>```
+
+## REST Paths (for the first version)
+
+### GET
+
+`/<database_name>/tables/<table_name>/rows`: Return a JSON of all data from the table <table_name>.
+`/<database_name>/tables/<table_name>/rows/<trid>`: Return a JSON of the data from
+the table <table_name> with TRID == <trid>.
+
+Example:
+
+```json
+...
+{
+   "col1":"ValCol1",
+   "col2":"ValCol2",
+   ...
+}
+```
+
+### POST
+
+`/databases/<database_name>/tables/<table_name>/rows`: Create a new row
+
+- Body: the name of the columns and their values: `{ "col1": "ValCol1", "col2": "ValCol2", ...}`
+- Update the row with TRID = <trid>
+- Return the Response protobuf converted into JSON.
+
+### PUT
+
+`/databases/<database_name>/tables/<table_name>/rows/<trid>`:
+
+- Body: the name of the columns and their values: `{ "col1": "ValCol1", "col2": "ValCol2", ...}`
+- Update the row with TRID = <trid>
+- Return the Response protobuf converted into JSON.
+
+### DELETE
+
+`/databases/<database_name>/tables/<table_name>/rows/<trid>`:
+
+- Body: none
+- Delete the row with TRID = <trid>
+- Return the Response protobuf converted into JSON.
+
+## Protocol
+
+### GET
+
+1. A client sends a query to the server:
+`curl --user "Siodb_user_name"  https://siodb-srv:80/<path>`
+2. `siodb_rest_mgr` received the request and query the IOMgr to see if the token is valid.
+    - If valid: it goes to step 3
+    - If not valid, it return the message `Status::CODE_401, "Unauthorized"`
+3. `siodb_rest_mgr` receive the request.
+4. `siodb_rest_mgr` convert the request into a `DatabaseEngineRequest` converting the query
+into a SQL query.
+5. `siodb_rest_mgr` send the command to the IOMgr.
+6. IOMgr does it work and send the `DatabaseEngineResponse` along with the raw data.
+7. `siodb_rest_mgr` writes into a JSON the response converted from `DatabaseEngineResponse`.
+8. `siodb_rest_mgr` receive raw data and form the resultset in a JSON form and append it
+the JSON to be responded.
+9. `siodb_rest_mgr` send the JSON response to the client.
+
+All values from IOMgr would be converted into UTF-8 JSON.
+
+### POST, PUT, DELETE
+
+1. A client send a query to the server: `curl --user "Siodb_user_name"  https://siodb-srv:80/<path>`
+2. `siodb_rest_mgr` received the request and query the IOMgr to see if the token is valid.
+    - If valid: it goes to step 3
+    - If not valid, it return the message `Status::CODE_401, "Unauthorized"`
+3. `siodb_rest_mgr` receive the request.
+4. `siodb_rest_mgr` convert the request into a `DatabaseEngineRequest` converting the query into
```

Comment:
There is a big problem. Because rest server doesn't know anything about data types of columns, it cannot protect from SQL injection. Only iomgr knows column data types. So JSON received from client must be sent to iomgr as is, with additional information what operation do we want: SELECT ALL/SELECT WITH TRID/UPDATE/INSERT/DELETE, and only iomgr can correctly extract data from json knowing expected data types.

Reply:
One question: currently, does the IOmgr protect us against SQL injection? Because the main point of having the prepared statements asap is to be protected from SQL injection. If the we are yet protected, no need to work on prepared statements now.

Reply:
No, iomgr does not protect us from SQL injection currently. It must be implemented via prepared statements for regular clients and via similar mechanism for JSON. Here I mean that the only place where you can implement protection is inside iomgr.